### PR TITLE
Bom workspace

### LINKED ISSUE #3656: Create a BOM for cargo workspace
We could define all dependencies version in our top level workspace cargo toml file. 
workspace:
```toml
[workspace.dependencies]
tonic = { version = "0.8.3", features = ["gzip"] }
tonic-build = "0.8.4"
```
modules:
```toml
tonic = { workspace = true }
tonic-build = { workspace = true }
```

Of course you could override the version but it’s cleaner to have only one version peer workspace.

If you are interested, I’ll open an issue to keep this.

https://doc.rust-lang.org/cargo/reference/workspaces.html#the-dependencies-table

### PR DESCRIPTION
**Pull request by litchipi**
Closes #3656

I created a Python script to migrate from `1 Cargo.toml per package` to `1 workspace Cargo.toml` (when it comes to dependencies)

When a dependency was not "compatible" with the workspace one (not exactly the same version for ex), I override the version to the one used in the workspace, but added a comment:
```
# BOM UPGRADE       Revert to "X" if problem
```

This way we can easily have a fix for anything breaking.

### CODE REVIEW

**Comment by litchipi** [comment]
(I'll delete the script I made just before merging, this way I can fix up things if needed)

**Comment by litchipi** [comment]
Note that this upgraded some versions on the project

I'll re-run the script on the latest `testnet` branch once it's ready, for a final review

File: Cargo.toml [hunk 1]
```diff
@@ -57,3 +57,126 @@ opt-level = 3 # Speed-up the CI
 #
 # * sandbox: for testing purpose, genesis timestamps is set as now + 9 seconds.
 # The saved snapshot can then be used to restart the network from the snapshot.
+[workspace.dependencies]
+aes-gcm = "0.10"
+anyhow = "1.0"
+async-trait = "0.1.58"
+atty = "0.2"
+bitvec = {version = "1.0"}
```
Comment (aoudiamoncef) [inline Cargo.toml:65]: this could simplified no ?

File: Cargo.toml [hunk 2]
```diff
@@ -57,3 +57,126 @@ opt-level = 3 # Speed-up the CI
...
+lazy_static = "1.4.0"
+lsmtree = "=0.1.1"
+machine = {git = "https://github.com/antifuchs/machine", "branch" = "fix-workspace-build"}
+massa-proto-rs = {git = "https://github.com/massalabs/massa-proto-rs", "rev" = "18ec02f"}
+massa-sc-runtime = {git = "https://github.com/massalabs/massa-sc-runtime", "branch" = "main"}
+massa_api = {path = "./massa-api"}
+massa_api_exports = {path = "./massa-api-exports"}
+massa_async_pool = {path = "./massa-async-pool"}
+massa_bootstrap = {path = "./massa-bootstrap"}
+massa_channel = {path = "./massa-channel"}
```
Comment (aoudiamoncef) [inline Cargo.toml:100]: Add comments or space to say that it's internal will be great

File: Cargo.toml [hunk 3]
```diff
@@ -57,3 +57,126 @@ opt-level = 3 # Speed-up the CI
...
+serde_qs = "0.11"
+serde_with = "2.1.0"
+serial_test = "1.0.0"
+sha2 = "0.10.6"
+socket2 = "0.4.7"
+stream_limiter = "3.2.0"
+structopt = {version = "0.3"}
+strum = "0.24"
+strum_macros = "0.24"
+substruct = {git = "https://github.com/sydhds/substruct"}
```
Comment (aoudiamoncef) [inline Cargo.toml:165]: using a rev could be safer

**Comment by litchipi** [comment]
Rebased over `testnet_25`, re-ran the script, and took your advices @aoudiamoncef
If CI is green and everything OK for you, let me know so we can merge this before someone modifies a `Cargo.toml` and creates a conflict :sweat_smile:

**Review by AurelienFT — APPROVED** [review]
FYI, this will break the code of Massembler (https://github.com/massalabs/massembler) this is not a problem just an information.

**Pull request merged**